### PR TITLE
Make classes final where applicable.

### LINKED ISSUE #3585: Mark eligible classes final
The issue was originally raised in https://github.com/doctrine/dbal/pull/3007#issuecomment-362630921.

A quick check with [Ocramius/Finalizer](https://github.com/Ocramius/Finalizer) shows:
```
$ finalizer finalizer:check-final-classes lib/
Following classes need to be made final:
+-----------------------------------------------------------------+
| Doctrine\DBAL\Sharding\ShardChoser\MultiTenantShardChoser       |
| Doctrine\DBAL\Sharding\SQLAzure\Schema\MultiTenantVisitor       |
| Doctrine\DBAL\Sharding\SQLAzure\SQLAzureFederationsSynchronizer |
| Doctrine\DBAL\Sharding\PoolingShardManager                      |
| Doctrine\DBAL\Id\TableGeneratorSchemaVisitor                    |
| Doctrine\DBAL\Driver\OCI8\Driver                                |
| Doctrine\DBAL\Driver\Mysqli\Driver                              |
| Doctrine\DBAL\Driver\Mysqli\MysqliStatement                     |
| Doctrine\DBAL\Driver\PDOSqlsrv\Driver                           |
| Doctrine\DBAL\Driver\PDOSqlsrv\Statement                        |
| Doctrine\DBAL\Driver\PDOMySql\Driver                            |
| Doctrine\DBAL\Driver\IBMDB2\DB2Connection                       |
| Doctrine\DBAL\Driver\IBMDB2\DB2Statement                        |
| Doctrine\DBAL\Driver\IBMDB2\DB2Driver                           |
| Doctrine\DBAL\Driver\SQLSrv\SQLSrvStatement                     |
| Doctrine\DBAL\Driver\SQLSrv\Driver                              |
| Doctrine\DBAL\Driver\SQLSrv\SQLSrvConnection                    |
| Doctrine\DBAL\Driver\SQLAnywhere\SQLAnywhereConnection          |
| Doctrine\DBAL\Driver\SQLAnywhere\Driver                         |
| Doctrine\DBAL\Driver\SQLAnywhere\SQLAnywhereStatement           |
| Doctrine\DBAL\Driver\PDOPgSql\Driver                            |
| Doctrine\DBAL\Driver\PDOOracle\Driver                           |
| Doctrine\DBAL\Driver\PDOSqlite\Driver                           |
| Doctrine\DBAL\Driver\StatementIterator                          |
| Doctrine\DBAL\Cache\ResultCacheStatement                        |
| Doctrine\DBAL\Cache\ArrayStatement                              |
| Doctrine\DBAL\Schema\Synchronizer\SingleDatabaseSynchronizer    |
| Doctrine\DBAL\Schema\Visitor\RemoveNamespacedAssets             |
| Doctrine\DBAL\Portability\Statement                             |
+-----------------------------------------------------------------+
```
For all finalized classes, all their properties and protected methods should be declard private and renamed if they violate the [PSR-2](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#42-properties) naming conventions.

### PR DESCRIPTION
**Pull request by jwage**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes
| Fixed issues | fixes #3585

### CODE REVIEW

File: lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php [hunk 1]
```diff
@@ -84,23 +84,23 @@ class MysqliStatement implements IteratorAggregate, Statement
...
-        $paramCount = $this->_stmt->param_count;
+        $paramCount = $this->stmt->param_count;
         if (0 >= $paramCount) {
             return;
         }
 
-        $this->types         = str_repeat('s', $paramCount);
-        $this->_bindedValues = array_fill(1, $paramCount, null);
+        $this->types        = str_repeat('s', $paramCount);
+        $this->bindedValues = array_fill(1, $paramCount, null);
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php]: Can we make it "bound" now? This should be the proper form AFAIK.

File: tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php [hunk 2]
```diff
@@ -79,27 +45,4 @@ private function assertIterationCallsFetchOncePerStep(Traversable $iterator, int
             $this->assertEquals($i + 1, $calls);
         }
     }
-
-    /**
-     * @return string[][]
-     */
-    public static function statementProvider() : iterable
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php]: It is fine to remove these tests. They enforce the lazy behavior in all statement classes however it's achieved by just using the statement iterator. As long as the statement iterator is instantiated internally by the implementations, you cannot really enforce it by a test w/o complicating the API. I believe not having these tests is a reasonable tradeoff.

File: lib/Doctrine/DBAL/Driver/PDOSqlite/Driver.php [hunk 3]
```diff
@@ -18,7 +18,7 @@
 class Driver extends AbstractSQLiteDriver
```
Comment (SenseException) [inline lib/Doctrine/DBAL/Driver/PDOSqlite/Driver.php]:
```suggestion
final class Driver extends AbstractSQLiteDriver
```

**Review by SenseException — CHANGES_REQUESTED** [review]

**Comment by morozov** [comment]
FWIW, the current state is:
```
$ finalizer finalizer:check-final-classes lib
Following classes need to be made final:
+--------------------------------------+
| Doctrine\DBAL\Driver\Mysqli\Driver   |
| Doctrine\DBAL\Driver\PDOPgSql\Driver |
+--------------------------------------+
```

**Comment by jwage** [comment]
@morozov fixed. Should we add that `finalizer` check to ci so it fails if a new class is introduced that should be final but isn't?

**Comment by morozov** [comment]
This is the first time I'm using it but it looks more suitable for development, not for CI. Specifically, it also [fails the build](https://github.com/Ocramius/Finalizer/blob/8f7bf97c3d6adb25af1f07d94f01c6c6a0a1ba84/src/Finalizer/Console/CheckFinalClassesCommand.php#L105) if it finds the classes mistakenly marked as final which is not what we need. I don't think we're introducing new classes that often that we need it enforced on CI.

**Comment by jwage** [comment]
I've got a few more tests to fix tomorrow.

File: lib/Doctrine/DBAL/Driver/PDOSqlite/Driver.php (hunk 3 above)
Comment (SenseException) [reply lib/Doctrine/DBAL/Driver/PDOSqlite/Driver.php]: This file is listed in `UPGRADE.md` as being final.

**Review by SenseException — DISMISSED** [review]

**Comment by morozov** [comment]
In addition to already finalized classes, the following driver-level classes are meant to be final but currently cannot be declared as such:

The following class have not been automatically finalized:

1. `MysqliConnection` because of its `getWrappedResourceHandle()` which is not used anywhere and is not part of any API convention (see #3534 for a similar feature request).
2. `OCI8Connection` because of its `getExecuteMode()` which is used by `OCI8Statement::execute()`.
3. `OCI8Statement` because of its _static_ `convertPositionalToNamedPlaceholders()` which is `public` only because it's tested directly which indicates violation of the SRP. It could be extracted to a utility class.
4. `PDOConnection` because it has child classes and the `createStatement()` method.
5. `PDOSqlsrv\Connection` because of its `createStatement()`.

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thank you @jwage :+1:

**Pull request merged**